src/util/include/util/raii/CLibrariesSPtr.h

```diff
@@ -0,0 +1,100 @@
+/*
+ * Xournal++
+ *
+ * RAII smart pointers for C library classes
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+#pragma once
+
+#include <utility>
+
+namespace xoj::util {
+/**
+ * @brief Simple template class for RAII smart pointer (aimed at Cairo/GTK/Poppler ref-counting classes)
+ * @param T The wrapper will store a pointer of type T
+ * @param H Handler class containing at least
+ *              static T* ref(T*):
+ *              static void unref(T*);
+ *              static T* adopt(T*);  // What to do to a pointer when we adopt an instance (typically the identity)
+ *          and optionally (for floating refs)
+ *              static T* ref_sink(T*);
+ */
+template <typename T, class H>
+class CLibrariesSPtr {
+public:
+    CLibrariesSPtr() = default;
+
+    using handler_type = H;
+
+private:
+    static auto safeRef(T* ptr) -> T* { return ptr ? H::ref(ptr) : ptr; }
+    static auto safeUnref(T* ptr) -> void {
+        if (ptr) {
+            H::unref(ptr);
+        }
+    }
+    static auto safeSinkRef(T* ptr) -> T* { return ptr ? H::sink_ref(ptr) : ptr; }
+    static auto safeReset(T*& ptr, T* val) -> void { safeUnref(std::exchange(ptr, val)); }
+
+public:
+    ~CLibrariesSPtr() { safeUnref(p); }
+
+    CLibrariesSPtr(const CLibrariesSPtr& other) { p = safeRef(other.p); }
+    CLibrariesSPtr(CLibrariesSPtr&& other) { p = std::exchange(other.p, nullptr); }
+
+    CLibrariesSPtr& operator=(const CLibrariesSPtr& other) {
+        if (this != &other) {
+            safeReset(p, safeRef(other.p));
+        }
+        return *this;
+    }
+    CLibrariesSPtr& operator=(CLibrariesSPtr&& other) {
+        if (this != &other) {
+            safeReset(p, std::exchange(other.p, nullptr));
+        }
+        return *this;
+    }
+
+    constexpr static struct Adopt {
+    } adopt = Adopt();
+    constexpr static struct Ref {
+    } ref = Ref();
+    constexpr static struct RefSink {
+    } refsink = RefSink();
+
+    CLibrariesSPtr(T* p, Adopt = adopt): p(H::adopt(p)) {}
```

Comment:
I think I found an acceptable solution: I added to the handler class H a function T* adopt(T*) which:
- For cairo objects: simply returns the input pointer
- For GObjects: tests if the ref is floating and sinks if necessary.

This makes both `GObjectSPtr p(g_object_new(...))` and `WidgetSPtr p(gtk_create_some_widget(....))` safe (note that g_object_new does not return a floating ref, while gtk's widget creating functions typically do -- with the notable exception of our own gtk_xournal_new).

Reply:
The thing is, that this check should not be necessary, because it is well documented, in which state the gobject is. I would only do an assert for debug builds.

Reply:
`g_object_take_ref` this is what we want :-) https://docs.gtk.org/gobject/method.Object.take_ref.html

Reply:
Except the linux CI does not know it exists... :-(

Reply:
> Except the linux CI does not know it exists... :-(

`g_object_take_ref` was introduced in glib 2.70. Ubuntu 18.04 LTS has glib 2.26 and ubuntu 20.04 has glib 2.64... So we can't rely on it. To bad!

Reply:
I went for using `g_object_take_ref` as soon as the glib version is high enough, with a fallback to the previous handmade "if floating sink, otherwise do nothing".

In the case of outdated glib2's, I would not replace the test by an assert for the following reason:
- this is the only way to actually get the exact same outcome as a call to `g_object_take_ref`
- us devs typically have an up-to-date library, and we would not see the assertion failures
- it offers more flexibility and safety